zrml/neo-swaps/src/lib.rs

```diff
@@ -0,0 +1,903 @@
+        pub fn deploy_pool(
+            origin: OriginFor<T>,
+            #[pallet::compact] market_id: MarketIdOf<T>,
+            #[pallet::compact] amount: BalanceOf<T>,
+            spot_prices: Vec<BalanceOf<T>>,
+            swap_fee: BalanceOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            let who = ensure_signed(origin)?;
+            let asset_count = T::MarketCommons::market(&market_id)?.outcomes() as u32;
+            ensure!(spot_prices.len() == asset_count as usize, Error::<T>::IncorrectVecLen);
+            Self::do_deploy_pool(who, market_id, amount, spot_prices, swap_fee)?;
+            Ok(Some(T::WeightInfo::deploy_pool(asset_count)).into())
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        #[require_transactional]
+        fn do_buy(
+            who: T::AccountId,
+            market_id: MarketIdOf<T>,
+            asset_out: AssetOf<T>,
+            amount_in: BalanceOf<T>,
+            min_amount_out: BalanceOf<T>,
+        ) -> DispatchResult {
+            ensure!(amount_in != Zero::zero(), Error::<T>::ZeroAmount);
+            let market = T::MarketCommons::market(&market_id)?;
+            ensure!(market.status == MarketStatus::Active, Error::<T>::MarketNotActive);
+            Self::try_mutate_pool(&market_id, |pool| {
+                ensure!(pool.contains(&asset_out), Error::<T>::AssetNotFound);
+                // Defensive check (shouldn't ever happen)!
+                ensure!(
+                    pool.calculate_spot_price(asset_out)? <= MAX_SPOT_PRICE.saturated_into(),
+                    Error::<T>::Unexpected
+                );
+                ensure!(amount_in <= pool.calculate_max_amount_in(), Error::<T>::NumericalLimits);
+                T::AssetManager::transfer(pool.collateral, &who, &pool.account_id, amount_in)?;
+                let FeeDistribution {
+                    remaining: amount_in_minus_fees,
+                    swap_fees: swap_fee_amount,
+                    external_fees: external_fee_amount,
+                } = Self::distribute_fees(market_id, pool, amount_in)?;
+                let swap_amount_out =
+                    pool.calculate_swap_amount_out_for_buy(asset_out, amount_in_minus_fees)?;
+                let amount_out = swap_amount_out.saturating_add(amount_in_minus_fees);
+                ensure!(amount_out >= min_amount_out, Error::<T>::AmountOutBelowMin);
+                // Instead of letting `who` buy the complete sets and then transfer almost all of
+                // the outcomes to the pool account, we prevent `(n-1)` storage reads by using the
+                // pool account to buy. Note that the fees are already in the pool at this point.
+                T::CompleteSetOperations::buy_complete_set(
+                    pool.account_id.clone(),
+                    market_id,
+                    amount_in_minus_fees,
+                )?;
+                T::AssetManager::transfer(asset_out, &pool.account_id, &who, amount_out)?;
```

Comment:
Based on the pdf doc you provided the user should receive `y(x)` after the buy, not `r_i + x`

```suggestion
                T::AssetManager::transfer(asset_out, &pool.account_id, &who, swap_amount_out)?;
```